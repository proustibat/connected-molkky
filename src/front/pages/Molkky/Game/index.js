import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import PositionChecker from '@components/PositionChecker';
import { DataContextProvider } from '@root/front/contexts/DataContext';
import { getRandomPosition } from '@utils/services';
import Menu from '@components/Menu';

const Game = ({ title }) => {
  const [positionData, setPositionData] = useState([]);

  useEffect(() => {
    // Get random data for skittles
    setPositionData(Array.from({ length: 12 }, (_, i) => ({
      value: i + 1,
      position: getRandomPosition(),
    })));
  }, []);

  return (
    <DataContextProvider value={positionData}>
      <Menu title={title} />
      <div className="section no-pad-bot">
        <div className="container">
          <PositionChecker />
        </div>
      </div>
    </DataContextProvider>
  );
};

Game.propTypes = {
  title: PropTypes.string,
};

Game.defaultProps = {
  title: null,
};

Game.displayName = 'Game';

export default Game;
