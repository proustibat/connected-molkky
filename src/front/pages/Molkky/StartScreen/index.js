import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import PositionChecker from '@components/PositionChecker';
import { DataContextProvider } from '@root/front/contexts/DataContext';
import { getRandomPositionData } from '@utils/services';
import Button from '@components/Button';
import Menu from '@components/Menu';
import CatSVG from '@root/front/svg/cat.svg';
import DogSVG from '@root/front/svg/dog.svg';
import TeamButton from '@components/TeamButton';

const StartScreen = ({ title }) => {
  const [positionData, setPositionData] = useState([]);
  const [startReady, setStartReady] = useState(false);
  const [refreshDataTimer, setRefreshDataTimer] = useState(null);
  const [selectedTeam, setSelectedTeam] = useState('cat');

  const createFakeServer = () => {
    // Get random data for skittles
    const timer = setInterval(() => {
      setPositionData(getRandomPositionData());
    }, 3000);
    setRefreshDataTimer(timer);
  };

  const destroyFakeServer = () => {
    clearInterval(refreshDataTimer);
    setRefreshDataTimer(null);
  };

  useEffect(() => {
    createFakeServer();
  }, []);

  const onStartClick = () => {
    console.log('START');
    destroyFakeServer();
  };

  const onPositionReadyChange = (value) => {
    setStartReady(value);
  };

  const onTeamSelect = (team) => {
    setSelectedTeam(team);
  };


  return (
    <DataContextProvider value={positionData}>
      <Menu title={title} />
      <div className="section">
        <div className="container">
          <h1>Players</h1>
          <p className="flow-text">Which team plays first?</p>
          <div style={{ display: 'flex', justifyContent: 'center' }}>
            <TeamButton icon={CatSVG} style={{ marginRight: '2rem' }} name="Team Cat" value="cat" onClick={onTeamSelect} selected={selectedTeam === 'cat'} />
            <TeamButton icon={DogSVG} style={{ marginLeft: '2rem' }} name="Team Dog" value="dog" onClick={onTeamSelect} selected={selectedTeam === 'dog'} />
          </div>
        </div>
      </div>
      <div className="section">
        <div className="container">
          <h1>Playground</h1>
          <PositionChecker
            onReadyChange={onPositionReadyChange}
          />
          <Button
            onClick={onStartClick}
            disabled={!startReady}
          >
            Play
          </Button>
        </div>
      </div>
    </DataContextProvider>
  );
};

StartScreen.propTypes = {
  title: PropTypes.string,
};

StartScreen.defaultProps = {
  title: null,
};

StartScreen.displayName = 'StartScreen';

export default StartScreen;
