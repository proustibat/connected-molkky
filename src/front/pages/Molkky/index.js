import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import gameConfig from '@root/gameConfig';
import socketIOClient from 'socket.io-client';

const Molkky = ({ title }) => {
  const [skittlesState, setSkittlesState] = useState({});
  const colStyle = { width: '25%' };

  useEffect(() => {
    const socket = socketIOClient(gameConfig.socketServer);
    socket.on('UPDATE', (data) => {
      setSkittlesState(data);
    });
  }, []);

  const renderSkittles = () => (
    <table className="striped">
      <thead>
        <tr>
          <th style={colStyle}>Mac</th>
          <th style={colStyle}>Value</th>
          <th style={colStyle}>Position</th>
          <th style={colStyle}>Battery</th>
        </tr>
      </thead>

      <tbody>
        {
          Object.entries(skittlesState)
            .sort(([, { value: valA }], [, { value: valB }]) => {
              let comparison = 0;
              if (valA > valB) {
                comparison = 1;
              } else if (valA < valB) {
                comparison = -1;
              }
              return comparison;
            }).map(([mac, { position, battery, value }]) => (
              <tr key={mac}>
                <td style={colStyle}>{mac}</td>
                <td style={colStyle}>{value}</td>
                <td style={colStyle}>{position}</td>
                <td style={colStyle}>{battery}</td>
              </tr>
            ))
        }
      </tbody>
    </table>
  );

  return (
    <div className="section no-pad-bot">
      <div className="container">
        {title && <h1>{ title }</h1>}
        {renderSkittles()}
      </div>
    </div>
  );
};

Molkky.propTypes = {
  title: PropTypes.string,
};

Molkky.defaultProps = {
  title: null,
};

Molkky.displayName = 'Molkky';

export default Molkky;
