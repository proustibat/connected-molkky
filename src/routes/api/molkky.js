import CurrentGame from '@root/CurrentGame';
import express from 'express';
import middleware from '@root/middleware';

const router = express.Router();

router.post('/start', middleware.checkConditionsToStartGame, async (req, res) => {
  const { body: { teams, playingTeam } } = req;
  // eslint-disable-next-line no-new
  new CurrentGame({ teams, playingTeam });
  res.json({
    scores: CurrentGame.instance.scores,
    currentTurn: CurrentGame.instance.currentTurn,
  }).end();
});

router.post('/score', middleware.checkGameStarted, middleware.checkTeam, middleware.checkPoints, async (req, res) => {
  const { body: { team, points } } = req;
  const currentGame = CurrentGame.instance;
  currentGame.addPoints({ team, points });
  res.json({
    scores: CurrentGame.instance.scores,
    currentTurn: CurrentGame.instance.currentTurn,
  }).end();
});

router.post('/miss', middleware.checkGameStarted, middleware.checkTeam, async (req, res) => {
  const currentGame = CurrentGame.instance;
  currentGame.missTurn();
  res.json({
    scores: CurrentGame.instance.scores,
    currentTurn: CurrentGame.instance.currentTurn,
  }).end();
});


module.exports = router;
