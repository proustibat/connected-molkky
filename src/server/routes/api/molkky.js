import CurrentGame from '@root/server/CurrentGame';
import express from 'express';
import middleware from '@root/server/middleware';

const router = express.Router();

router.post('/start', middleware.checkConditionsToStartGame, async (req, res) => {
  const { body: { teams, playingTeam } } = req;
  // eslint-disable-next-line no-new
  const currentGame = new CurrentGame({ teams, playingTeam });
  res.json({
    scores: currentGame.scores,
    currentTurn: currentGame.currentTurn,
  }).end();
});

router.post('/score', middleware.checkGameStarted, middleware.checkTeam, middleware.checkPoints, async (req, res) => {
  const { body: { team, points } } = req;
  const currentGame = CurrentGame.instance;
  currentGame.addPoints({ team, points });
  res.json({
    scores: currentGame.scores,
    currentTurn: currentGame.currentTurn,
  }).end();
});

router.post('/miss', middleware.checkGameStarted, middleware.checkTeam, async (req, res) => {
  const currentGame = CurrentGame.instance;
  currentGame.missTurn();
  res.json({
    scores: currentGame.scores,
    currentTurn: currentGame.currentTurn,
  }).end();
});

router.post('/reset', async (req, res) => {
  const currentGame = CurrentGame.instance;
  currentGame.reset();
  res.json({
    scores: null,
    currentTurn: null,
  }).end();
});


module.exports = router;
