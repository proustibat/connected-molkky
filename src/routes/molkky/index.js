import express from 'express';
import React from 'react';
import Molkky from '@pages/Molkky';
import Game from '@pages/Molkky/Game';
import renderTemplate from '@root/renderTemplate';

const router = express.Router();

/* GET hue page. */
router.get('/', (req, res) => {
  const data = { title: 'Molkky' };
  res.writeHead(200, { 'Content-Type': 'text/html' });
  res.end(renderTemplate(<Molkky {...data} />));
});

router.get('/game', (req, res) => {
  const data = { title: 'Molkky Game' };
  res.writeHead(200, { 'Content-Type': 'text/html' });
  res.end(renderTemplate(<Game {...data} />));
});


module.exports = router;
