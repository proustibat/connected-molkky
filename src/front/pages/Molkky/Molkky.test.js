import { cleanup, render } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import Molkky from './index';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { dataFromSocketUPDATE } from '@fixtures/molkky';
import socketIOClient from 'socket.io-client';

jest.mock('socket.io-client', () => {
  const emit = jest.fn();
  const on = jest.fn();
  const socket = { emit, on };
  return jest.fn(() => socket);
});

const givenProps = {
  title: 'Yo',
};

describe('Molkky', () => {
  afterEach(cleanup);
  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should render the component correctly with no skittles data', () => {
    // Given / When
    const component = shallow(<Molkky {...givenProps} />);

    // Then
    expect(component).toHaveLength(1);
    expect(component.find('h1')).toHaveLength(1);
    expect(component.find('h1').first().text()).toBe(givenProps.title);
    expect(component.find('table')).toHaveLength(1);
    expect(component.find('thead')).toHaveLength(1);
    expect(component.find('tr')).toHaveLength(1);
    expect(component.find('th')).toHaveLength(4);
  });

  it('should connect to the socket and listen UPDATE event', () => {
    // Given / When
    mount(<Molkky {...givenProps} />);

    // Then
    expect(socketIOClient).toHaveBeenCalledTimes(1);
    expect(socketIOClient).toHaveBeenLastCalledWith('localhost:8888');
    expect(socketIOClient().on).toHaveBeenCalledTimes(1);
    expect(socketIOClient().on.mock.calls[0][0]).toBe('UPDATE');
  });

  it('should use data from UPDATE socket event', async () => {
    // Given
    let component;
    act(() => { component = render(<Molkky {...givenProps} />); });

    // When
    act(() => { socketIOClient().on.mock.calls[0][1](dataFromSocketUPDATE); });

    // Then
    expect(component.container).toMatchSnapshot();
  });
});
