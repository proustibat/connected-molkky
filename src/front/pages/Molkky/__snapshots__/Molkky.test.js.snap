// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molkky should use data from UPDATE socket event 1`] = `
<div>
  <div
    class="section no-pad-bot"
  >
    <div
      class="container"
    >
      <h1>
        Yo
      </h1>
      <table
        class="striped"
      >
        <thead>
          <tr>
            <th>
              Mac
            </th>
            <th>
              Position
            </th>
            <th>
              Battery
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              123qwe123
            </td>
            <td>
              UPRIGHT
            </td>
            <td>
              99
            </td>
          </tr>
          <tr>
            <td>
              456rty789
            </td>
            <td>
              KNOCKED_OVER
            </td>
            <td>
              43
            </td>
          </tr>
          <tr>
            <td>
              102ajd647
            </td>
            <td />
            <td>
              11
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
